dali/test/python/test_RNNT_data_pipeline.py

```diff
@@ -0,0 +1,155 @@
+# Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import nvidia.dali
+import nvidia.dali.ops as ops
+import nvidia.dali.types as types
+import numpy as np
+import os
+import test_utils
+
+dali_extra_path = test_utils.get_dali_extra_path()
+
+
+class RnntTrainPipeline(nvidia.dali.pipeline.Pipeline):
+    def __init__(self,
+                 device_id,
+                 n_devices,
+                 file_root,
+                 file_list,
+                 batch_size,
+                 sample_rate=16000,
+                 window_size=.02,
+                 window_stride=.01,
+                 nfeatures=64,
+                 nfft=512,
+                 frame_splicing_factor=3,
+                 silence_threshold=-80,
+                 preemph_coeff=.97,
+                 lowfreq=0.0,
+                 highfreq=0.0,
+                 num_threads=1, exec_async=False, exec_pipelined=False):
+        super().__init__(batch_size, num_threads, device_id,
+                         exec_async=exec_async,
+                         exec_pipelined=exec_pipelined, seed=42)
+
+        self.sample_rate = sample_rate
+        self.frame_splicing_factor = frame_splicing_factor
+
+        self.read = ops.FileReader(file_root=file_root, file_list=file_list, device="cpu",
+                                   shard_id=device_id, num_shards=n_devices)
+
+        self.decode = ops.AudioDecoder(device="cpu", dtype=types.FLOAT, downmix=True)
+
+        self.normal_distribution = ops.NormalDistribution(device="cpu")
+
+        self.preemph = ops.PreemphasisFilter(preemph_coeff=preemph_coeff)
+
+        self.spectrogram = ops.Spectrogram(device="cpu", nfft=nfft,
+                                           window_length=window_size * sample_rate,
+                                           window_step=window_stride * sample_rate)
+
+        self.mel_fbank = ops.MelFilterBank(device="cpu", sample_rate=sample_rate, nfilter=nfeatures,
+                                           normalize=True, freq_low=lowfreq, freq_high=highfreq)
+
+        self.log_features = ops.ToDecibels(device="cpu", multiplier=np.log(10), reference=1.0,
+                                           cutoff_db=-80)
+
+        self.get_shape = ops.Shapes(device="cpu")
+
+        self.normalize = ops.Normalize(axes=[0], device="cpu")
```

Comment:
This implies per-feature normalization. Shouldn't it be configurable, as it seemingly is in the original pipeline?

Reply:
In offline discussion we decided to go without configuration of this one